### PR TITLE
address multiple timestamp issues with multi

### PR DESCRIPTION
**Pull request by notimaginative**
A bit overkill for some of this but just covering the bases.

 * fix issues with in-game join
 * fix multi messaging and multi pause screen
 * make sure maintenance stuff (pings, voice, file xfer, etc.) isn't
   affected by pausing or time compression
 * fix sync issues caused by options menu and screenshots

### CODE REVIEW

**Comment by notimaginative** [comment]
@Goober5000 Please check me on this when you get a chance. Some of it probably isn't needed in the general UI but I preferred to go overkill rather than miss something.

File: code/network/multi_ingame.cpp [hunk 1]
```diff
@@ -944,13 +944,13 @@ void multi_ingame_handle_timeout()
...
 	// if we've timed out, leave the game
-	if( timestamp_elapsed(Ingame_time_left) ) {
+	if( ui_timestamp_elapsed(Ingame_time_left) ) {
 		multi_quit_game(PROMPT_NONE, MULTI_END_NOTIFY_INGAME_TIMEOUT, MULTI_END_ERROR_NONE);
 		return;
 	}
 
 	// otherwise, blit how much time we have left
-	int time_left = timestamp_until(Ingame_time_left) / 1000;
+	int time_left = ui_timestamp_until(Ingame_time_left) / 1000;
```
Comment (Goober5000) [inline code/network/multi_ingame.cpp:953]: I recommend using `MILLISECONDS_PER_SECOND` here.

File: code/network/multi_ingame.cpp [hunk 2]
```diff
@@ -475,7 +475,7 @@ static int Multi_ingame_timer_coords[GR_NUM_RESOLUTIONS][2] = {
 //#define MULTI_INGAME_TIME_LEFT_Y			411
 
 #define MULTI_INGAME_TIME_SECONDS		(1000 * 15)
```
Comment (Goober5000) [inline code/network/multi_ingame.cpp:477]: I recommend setting this to just 15 and then using `MILLISECONDS_PER_SECOND` when the actual timestamp is created.

File: code/network/multi.cpp [hunk 3]
```diff
@@ -1299,11 +1299,11 @@ void multi_do_frame()
 		int idx;
 		for(idx=0;idx<MAX_PLAYERS;idx++){
 			if(MULTI_CONNECTED(Net_players[idx]) && (Net_player != &Net_players[idx])){
-				if((Multi_client_update_times[idx] < 0) || timestamp_elapsed_safe(Multi_client_update_times[idx], 1000)){
+				if ( !Multi_client_update_times[idx].isValid() || ui_timestamp_elapsed_safe(Multi_client_update_times[idx], 1000) ) {
```
Comment (Goober5000) [inline code/network/multi.cpp:1302]: Might want to use `MILLISECONDS_PER_SECOND` here.

File: code/network/multi.cpp [hunk 4]
```diff
@@ -1399,11 +1399,11 @@ void multi_pause_do_frame()
 
 		for(idx=0;idx<MAX_PLAYERS;idx++){
 			if(MULTI_CONNECTED(Net_players[idx]) && (Net_player != &Net_players[idx])){			
-				if((Multi_client_update_times[idx] < 0) || timestamp_elapsed_safe(Multi_client_update_times[idx], 1000)){
+				if ( !Multi_client_update_times[idx].isValid() || ui_timestamp_elapsed_safe(Multi_client_update_times[idx], 1000) ) {
```
Comment (Goober5000) [inline code/network/multi.cpp:1402]: Here too

File: code/network/multi_observer.cpp [hunk 5]
```diff
@@ -59,7 +59,7 @@ void multi_obs_create_player(int player_num,char *name,net_addr *addr,player *pl
 	Net_players[player_num].last_heard_time = timer_get_fixed_seconds();
 	Net_players[player_num].reliable_socket = PSNET_INVALID_SOCKET;
 	Net_players[player_num].s_info.kick_timestamp = -1;
-	Net_players[player_num].s_info.voice_token_timestamp = -1;
+	Net_players[player_num].s_info.voice_token_timestamp = UI_TIMESTAMP::invalid();
 	Net_players[player_num].s_info.player_collision_timestamp = timestamp(0);
```
Comment (Goober5000) [inline code/network/multi_observer.cpp:63]: Looks like you can convert this to use `TIMESTAMP`, and use `TIMESTAMP::immediate()`

File: code/network/multi_observer.cpp [hunk 6]
```diff
@@ -59,7 +59,7 @@ void multi_obs_create_player(int player_num,char *name,net_addr *addr,player *pl
 	Net_players[player_num].last_heard_time = timer_get_fixed_seconds();
```
Comment (Goober5000) [inline code/network/multi_observer.cpp:59]: Looks like `last_heard_time` could be converted to `TIMESTAMP` or `UI_TIMESTAMP`

File: code/network/multi_observer.cpp [hunk 7]
```diff
@@ -59,7 +59,7 @@ void multi_obs_create_player(int player_num,char *name,net_addr *addr,player *pl
...
 	Net_players[player_num].reliable_socket = PSNET_INVALID_SOCKET;
 	Net_players[player_num].s_info.kick_timestamp = -1;
```
Comment (Goober5000) [inline code/network/multi_observer.cpp:61]: I'd say `kick_timestamp` could be converted too

File: code/network/multiutil.cpp [hunk 8]
```diff
@@ -800,7 +800,7 @@ void multi_create_player( int net_player_num, player *pl, const char* name, net_
 	Net_players[net_player_num].last_heard_time = timer_get_fixed_seconds();
 	Net_players[net_player_num].reliable_socket = PSNET_INVALID_SOCKET;
 	Net_players[net_player_num].s_info.kick_timestamp = -1;
-	Net_players[net_player_num].s_info.voice_token_timestamp = -1;
+	Net_players[net_player_num].s_info.voice_token_timestamp = UI_TIMESTAMP::invalid();
 	Net_players[net_player_num].s_info.player_collision_timestamp = timestamp(0);
 	Net_players[net_player_num].s_info.tracker_security_last = -1;
 	Net_players[net_player_num].s_info.target_objnum = -1;
```
Comment (Goober5000) [inline code/network/multiutil.cpp:806]: Same comments re `kick_timestamp` and `player_collision_timestamp`

File: code/network/multiutil.cpp [hunk 9]
```diff
@@ -1724,7 +1725,7 @@ void multi_free_active_games()
 			backup = moveup;
 			moveup = moveup->next;
 			
-			vm_free(backup);
+			delete backup;
```
Comment (Goober5000) [inline code/network/multiutil.cpp:1728]: This function looks like a candidate for smart pointers, but that's out of scope for this PR.

File: code/network/multi_pxo.cpp [hunk 10]
```diff
@@ -561,7 +561,7 @@ int Multi_pxo_chat_count = 0;
 
 // extra delay time when switching channels
 #define MULTI_PXO_SWITCH_DELAY_TIME			2000
```
Comment (Goober5000) [inline code/network/multi_pxo.cpp:563]: I'd suggest making this `2 * MILLISECONDS_PER_SECOND`

File: code/network/multi_pxo.cpp [hunk 11]
```diff
@@ -1017,7 +1017,7 @@ void multi_pxo_help_button_pressed(int n);
 InetGetFile *Multi_pxo_ban_get = NULL;
 
 #define PXO_BANNER_SWAP_TIME			120000
```
Comment (Goober5000) [inline code/network/multi_pxo.cpp:1019]: I'd suggest making this `120 * MILLISECONDS_PER_SECOND`

File: code/network/multi_voice.cpp [hunk 12]
```diff
@@ -98,7 +98,7 @@ char *Multi_voice_playback_buffer = NULL;								// buffer for processing the ac
 // voice algorithm stuff
 // it would probably be good to base the timeout time on some multiple of our average ping to the server
 #define MV_ALG_TIMEOUT	500												// if start get new data for a window then a pause this long, play the window
-int Multi_voice_stamps[MULTI_VOICE_MAX_STREAMS];
+UI_TIMESTAMP Multi_voice_stamps[MULTI_VOICE_MAX_STREAMS];
 
 // NOTE : this should be > then MULTI_VOICE_MAX_TIME + the time for the data to come over a network connection!!
 #define MULTI_VOICE_TOKEN_TIMEOUT					7000				// timeout - server will take the token back if he does not hear from the guy in this amount of time
```
Comment (Goober5000) [inline code/network/multi_voice.cpp:104]: Suggest `7 * MILLISECONDS_PER_SECOND`

File: code/network/multi_voice.cpp [hunk 13]
```diff
@@ -146,7 +146,7 @@ int Multi_voice_player_prefs[MAX_PLAYERS];							// player bitflag preferences
 
 // voice status data - used for determing the result of multi_voice_status
 #define MULTI_VOICE_DENIED_TIME						1000				// how long to display the "denied" status
```
Comment (Goober5000) [inline code/network/multi_voice.cpp:148]: Suggest `1 * MILLISECONDS_PER_SECOND`

File: code/network/multi_voice.cpp [hunk 14]
```diff
@@ -2076,7 +2076,7 @@ void multi_voice_test_process_next_chunk()
 void multi_voice_test_record_start()
 {
 	// if there is test recording going on already, don't do anything
-	if(Multi_voice_test_record_stamp != -1){
+	if ( !Multi_voice_test_record_stamp.isValid() ) {
```
Comment (Goober5000) [inline code/network/multi_voice.cpp:2079]: I don't think the logical NOT is correct here.

**Review by Goober5000 — COMMENTED** [review]
I like this; these are good and sensible changes.  I don't think it's overkill at all.  I had several comments.

(The `MILLISECONDS_PER_SECOND` suggestions are optional; they might improve clarity but they might also get in the way.  Your call.  And I'm sure there are others I didn't catch.)

File: code/network/multi_voice.cpp (hunk 14 above)
Comment (notimaginative) [reply code/network/multi_voice.cpp:2079]: Yep, I got that wrong in a few places but must have missed this one when going back to fix them.

**Comment by notimaginative** [comment]
I think that covers the required changes. I make a couple of the `MILLISECONDS_PER_SECOND` changes, but I'd prefer to hold off on the rest for a proper cleanup PR and leave this one to focus on the real fixes.

File: code/network/multi_observer.cpp (hunk 7 above)
Comment (notimaginative) [reply code/network/multi_observer.cpp]: Done

File: code/network/multi.cpp (hunk 4 above)
Comment (notimaginative) [reply code/network/multi.cpp]: Fixed

File: code/network/multi.cpp (hunk 3 above)
Comment (notimaginative) [reply code/network/multi.cpp]: Fixed

File: code/network/multi_observer.cpp (hunk 6 above)
Comment (Goober5000) [reply code/network/multi_observer.cpp:59]: Ok.

**Review by Goober5000 — APPROVED** [review]

**Pull request merged**